### ISSUE TITLE
print header in csv file

### ISSUE THREAD

**Issue by mauricioaniche** [mauricioaniche/metricminer2#8]
should find a nice way to print the header of the csv

**Comment by mauricioaniche** [comment]
if you pass the header to the constructor, then you can validate any write() command with wrong number of parameters

**Comment by chripard** [comment]
hi @mauricioaniche , can you please be more specific on this?

**Comment by mauricioaniche** [comment]
Currently, it's up to the developer to print the header of the CSV. IF s/he doesn't do that, then the CSV ends up with no header.

My suggestion is in the `new CSVWriter(file)`, to add one more parameter, e.g.:

```
String[] header = { "col1", "col2", "col3" };

new CSVWriter("file.csv", header);
```

Then, `CSVWriter` writes this header as first line of the file. 

As know the class also knows the number of columns, if there's an invocation to `.write()` with the wrong number of columns, it should throw a `RepoDriller` exception explaining that!

By the way, this class is getting more complex. I'd appreciate if you submit an automated test for it. If you don't know how to do it, let me know, and I'll do it afterwards.

Did you get it?

**Comment by chripard** [comment]
Yes, i will have a look on this.

**Comment by chripard** [comment]
@mauricioaniche one question, the exception will be thrown from the CSVFile class right? The developer
should deal with it in main?

**Comment by chripard** [comment]
@mauricioaniche one question, the exception will be thrown from the CSVFile class right ? The developer
should deal with it in main ?

**Comment by mauricioaniche** [comment]
@chripard, yes, should be thrown from there!

I am actually thinking that, if that happens, we should stop RepoDriller. It's much better to fail fast in this situation, as the output will be unexpected. So, maybe you should:

- Make the current constructor of `CSVFile` deprecated.
- In the new constructor, ask for a `String[]`, which are the columns. Print them already.
- If the user passed the `String[]` in the new constructor, Throw a java `IllegalFormatException` (or you can create a new `IllegalCSVFormatException`) from the `CSVFile`, if values do not match columns.
- Catch this exception at the `RepositoryMining: processChangeSet`. It catches `Exception`, but add one more. If that happens, `log.fatal` the exception, and `System.exit(-1)`.

What do you think?

**Comment by chripard** [comment]
I think it is a great idea.

**Comment by nascimentolwtn** [comment]
I agree that if a user wants a header, it should be validated and fail if some line doesn't match the given number of columns.
Fail fast is perfect, but what about `System.exit(-1)` in the case of multiple repositories? All other running mining would be aborted. It already happened with me when I didn't catch RuntimeExceptions by myself. Or this case could be left to be dealt by user also.
In another hand, in my opinion the header could be somehow optional. As I'm looking to #19 and passing `PersistenceMechanism` to callback methods, the visitor could have more flexibility to write headers/footers, right?

**Comment by chripard** [comment]
I also think that the header should be optional.

**Comment by mauricioaniche** [comment]
Hi @nascimentolwtn, thanks for joining the discussion!

We can make it optional, yes. But, if you provide it, I don't see why it shouldn't fail if the writer doesn't match the columns. Can you see a case in which the number of values written doesn't match the required one? It seems always a bug to me.

Even if you have multiple repositories, I guess the output in the CSV has always the same structure, right?

**Comment by nascimentolwtn** [comment]
Yes, it's right that all should have the same structure.

But for some reason (eg writing Commit messages or other field that
contains comma and invalidate the column number), only that repository may
be faulty, in that case we should fail only that repo, and not all study.

In a case of a bug, all repos will naturally fail.

Does it make sense?

**Comment by mauricioaniche** [comment]
If this particular case happens, we should fix it at RepoDriller, right? (which is already fixed actually, tks to @chripard). I still don't see a case where you can have a valid `writer.write()` to one project and an invalid one to another project! :/

**Comment by nascimentolwtn** [comment]
@mauricioaniche , you're right. If a invalid CSV line is produced not due a faulty code, but due an untreated condition of a particular repository, it can also be considered a bug and then we could fail.

But if CSV validation could be optional, in cases that is worthy not to stop RepoDriller execution, user can let other projects continue to be processed, later he analyze that particular repository that produced the invalid CSV and then reprocess it.

For instance, In order to prevent a "invalid" CSV line, my visitor is removing line feeds and comma characters from using the regex `commitMessage.replaceAll("[\\r\\n,]+", "")`. By the way, a regex like this could be suggested by a CSVFile method in cases like this.
